### DISCUSSION TITLE
Custom component: complex impedance

### DISCUSSION THREAD

**Discussion by hahschaa** [SpiceSharp/SpiceSharp#173] [Q&A] [answered]
Hi Sven,

I switched from my own MNA solver to Spice# for analyzing the AC behavior of circuits containing only RCL components. Spice# is just so much faster and easier to use (So thank you for that 😄). However, one of the first things that I did with my own solver was combining multiple series and parallel components, of which I wasn't interested in their individual impact, into a single (frequency dependent) complex impedance. So, I was wondering if it was possible to do the same thing using Spice#.

I was delighted to read that it is possible to implement custom models and that there were some examples how to do it. However, after reading the custom components tutorial I was still totally lost at where to start implementing it. I checked out the source code to see if I could sort of reverse engineer how you implemented the RLC components, but to no avail; the auto-generated source code makes it really hard to see what's happening.

Do you think it is possible to implement this? Or is it a stupid idea to begin with? If not, can you maybe give me some pointers (or example code) to show how I should implement it? 

I guess it should be something like this, but I have no idea how to go further from here:
```csharp
[Pin(0, "Z+")]
[Pin(1, "Z-")]
public class Impedance : Component
{
    [ParameterName("pincount")] [ParameterInfo("Number of pins")]
    public const int ImpedancePinCount = 2;

    public Impedance(string name, string pos, string neg, Func<double, Complex> impedance)
        : base(name, ImpedancePinCount)
    {
        // ????
        Connect(pos, neg);
    }

    public override void CreateBehaviors(ISimulation simulation)
    {
        // ???
    }
}
```

Thank you in advance 😃

**Comment by svenboulanger** [comment, marked as answer]
Hi! I can imagine that what you want to do is a bit difficult especially knowing the documentation is lagging behind.

So, if you use the Spice# source code generator, it would take off some boilerplate code that would usually be pretty repetitive. However, in some cases I can't use the "standard" code, and so it is possible to still write it yourself. An example of such a case is subcircuits in Spice# (example [here](https://github.com/SpiceSharp/SpiceSharp/blob/master/SpiceSharp/Components/Subcircuits/Subcircuit.cs)).

Every entity likely needs 4 things:
- The parameters (a class that extends `ParameterSet<T>` where T is used for `ICloneable<T>`), like for the [resistor](https://github.com/SpiceSharp/SpiceSharp/blob/master/SpiceSharp/Components/RLC/Resistors/Parameters.cs).
- A binding context (a class that extends `IComponentBindingContext` that will be used to create simulation behaviors. In your case, you can reuse `SpiceSharp.Components.ComponentBindingContext` ([source](https://github.com/SpiceSharp/SpiceSharp/blob/master/SpiceSharp/Components/ComponentBindingContext.cs)).
- The entity/component, which is what your source code is. This class is responsible for creating behaviors for a simulation.
- The behaviors. In your case, you will need at least an `IBiasingBehavior` and an `IFrequencyBehavior`. They can be the same object, or not, just make sure there is only one object for each interface given to the simulation in `CreateBehaviors`.

### The parameters
```csharp
public class MyParameters : ParameterSet<MyParameters>
{
    // Your parameters and possibly an override for the cloning behavior (by default it just copies all fields)
}
```

### The component
```csharp
// This generic implementation of Component<T> automatically makes the "Parameter" property for you!
public class MyComponent : Component<MyParameters>
{
    public MyComponent(string name) : base(name, 2)
    {
        // You can connect already using Connect(), but you don't have to
        // You can also already initialize some parameters if you want
    }

    public override void CreateBehaviors(ISimulation simulation)
    {
        var behaviors = new BehaviorContainer(Name);
        var context = new ComponentBindingContext(this, simulation, behaviors);

        // This just creates behaviors if there isn't one already of the specified type
        behaviors.Build(simulation, context)
            .AddIfNo<IFrequencyBehavior>(context => new FrequencyBehavior(context))
            .AddIfNo<IBiasingBehavior>(context => new BiasingBehavior(context));
        simulation.EntityBehaviors.Add(behaviors);
    }
}
```

### Behaviors

```csharp
public class BiasingBehavior : Behavior, IBiasingBehavior
{
    private readonly Element<double> _diagonal, _rhs;

    public BiasingBehavior(IComponentBindingContext context)
    {
        // Get the parameters
        var parameters = context.GetParameters<Parameters>();

        // The state has some variables (some that are shared, some private), so we want to get those first
        var state = context.GetState<IBiasingSimulationState>();
        var nodeVariable = state.GetSharedVariable(context.Nodes[0]); // You can do this for all your other variables

        // Then we need to know which matrix and RHS row/columns that we can use, and for that we can use
        int nodeIndex = state.Map[nodeVariable];

        // Finally we can create matrix elements and right-hand-side vector elements with this
        _diagonal = state.Solver.GetElement(new MatrixLocation(nodeIndex, nodeIndex)); // Gets matrix element
        _rhs = state.Solver.GetElement(nodeIndex); // Gets RHS vector element, since you have passive linear components, you probably won't need RHS elements
    }

    public void Load()
    {
        // This is called every iteration of the biasing (obviously you would need to calculate the MNA contributions).
        _diagonal.Add(1.0);
        _rhs.Add(1.0);
    }
}
```

The frequency behavior is analog to the biasing behavior, except that you'll have to use the solver for complex elements. For that you need to use the `IComplexSimulationState` simulation state instead of the biasing one.

Hope this gets you more on the way...

**Reply by hahschaa** [reply]
Hi Sven. First of all thank you very much for your detailed answer. I think I almost got it working. At least, that's what I think since the simulation in running without any exceptions and I can calculate the correct impedance in the `Load(..)` method by doing something like:
```c#
var freq = this.state.Laplace.Imaginary / (2.0 * Math.PI);
var impedance = this.parameters.Impedance.Value?.Invoke(freq);
```

However, somehow I can't get the `ComplexPropertyExport` to export my complex current. Is there something I need to initialize here as well?

Thank you very much.

**Reply by svenboulanger** [reply]
Hi, happy to hear it's working out!

To use a property exporter, you need to override the [CreatePropertyGetter](https://github.com/SpiceSharp/SpiceSharp/blob/dace7bddaadf7ac22bc32abd2dc61b4d28afb13d/SpiceSharp/Parameters/ParameterSet.cs#L90) (defined in the `IParameterSet` interface) for your `IFrequencyBehavior`. If you use the default implementation, then you can also simply implement the `IExportPropertySet<P>` interface directly. Or, if that is annoying (especially if you have many properties), you can simply include the Spice#.Generator project from NuGet for your project and have the interface be generated automatically by the code generator.

If you decide to use the code generator, the total work should reduce to adding the `GeneratedParameters` attribute to your frequency behavior, and marking your properties with the `ParameterName` attribute (which then define the name of the property). The code generator will do the rest for you.

**Reply by hahschaa** [reply]
Thank you very much. Works like a charm :)
